README.org

```diff
@@ -1112,6 +1112,82 @@ Please provide the necessary important information with your bug report:
 When evaluating Consult-related code snippets you should enable [[https://www.gnu.org/software/emacs/manual/html_node/elisp/Lexical-Binding.html][lexical binding]].
 Consult often relies on lambdas and lexical closures.
 
+* Hacking
+** Creating a new asynchronous completion source
+
+If you have a completion source that's both dynamic and expensive to generate,
+=completing-read= may not be the best choice. Instead, =consult--read= serves as a
+thin wrapper around =completing-read= that provides this functionality.
+
+For example, consider the following script that splits its input on space:
+
+#+begin_src sh
+  #!/usr/bin/env bash
+
+  # simulate work
+  sleep .1
+
+  # generate completion candidates
+  printf "%s\n" "$*" | tr " " "\n" | sort
+#+end_src
+
+Let's assume this script is callable as =testibus hello world=.
+
+To have Consult use it as a completion source, use =consult--async-command=:
+
+#+begin_src emacs-lisp
+  (consult--read
+   (consult--async-command (lambda (input) (list "testibus" (s-trim input))))
+   :prompt "run testibus: ")
+#+end_src
+
+*** ...with live preview
+
+Implementing live preview is a little more involved and requires the definition
+of a =STATE= function as defined by =consult--with-preview=. The
+=consult--define-state= macro can assist by defining a function suitable to
+provide to =consult--read='s =:state= argument.
+
+First, define a 'preview' function with a name using the format
+=consult--%s-preview=. When called with no arguments, it should return a suitable
+'state' function:
+
+#+begin_src emacs-lisp
+  (defun consult--testibus-preview ()
```

Comment:
What I don't necessarily like about my example is that it requires defining symbols in the `consult` namespace. What are your thoughts here?

Reply:
In order to avoid the namespacing problem, do not use `consult--define-state`. For a hacking section it is better to be more explicit anyway, avoiding both `consult--define-state` and `consult--state-with-return`.

Reply:
I think I've done this correctly – at least, 'it works on my machine'. Let me know if I've missed the mark here.